### ISSUE TITLE
[Promote a Blog Post] Conversion tracking with Google Analytics

### ISSUE THREAD

**Issue by paagla123** [coa0329/BhaangWeb#65]
_From @daver78 on August 30, 2017 14:12_

@paagla123: Where are we with this?  

Who is assigned to this task?

Is it intentionally left off the P17 list?

_Copied from original issue: coa0329/Bhaang#42_

**Comment by paagla123** [comment]
@daver78 yes. We will deal with this after P17

**Comment by paagla123** [comment]
_From @daver78 on August 30, 2017 18:8_

@paagla123 OK, so should we leave it in the UI?  I can see pros & cons to that.

Who is assigned to GA research?

**Comment by paagla123** [comment]
@daver78 I may be confusing things here, but what's the difference between this and the new issue #47?

**Comment by paagla123** [comment]
_From @daver78 on August 31, 2017 4:17_

@paagla123 I see 3 discrete things:

1) One issue is making conversion tracking work (with Google, perhaps later with our own web tracking).  As I understand now, we don't intend for this to work before P17, but I'd like to know who will work on it.

2) Another issue is appending a UTM string, which is useful even if we're not doing conversion tracking.  Marketers like their UTM tagging.

3) A third issue is if we are not going to have conversion tracking for "Promote a Blog Post" ready by P17, should we hide that section in the UI.

What say you, good sir?

**Comment by paagla123** [comment]
@daver78 we're addressing issue#47 for P17 - that should cover #2 above. I don;t think we need to hide anything - it will be a good talking point.

**Comment by paagla123** [comment]
_From @daver78 on August 31, 2017 12:53_

@paagla123 Wonderful!

**Comment by paagla123** [comment]
_From @daver78 on September 13, 2017 13:33_

@paagla123 Where are we with this?  @paweenwich who is assigned to implement this feature?

**Comment by paagla123** [comment]
_From @paweenwich on September 14, 2017 1:48_

@daver78 Sorry, I have no idea what you gays talking about.  I know a bit about GA but not the Conversion tracking.   Could you explain the idea and give me a pointer?

**Comment by paagla123** [comment]
_From @daver78 on September 14, 2017 2:6_

@paagla123 I think you covered this with @paweenwich when you were in BKK?

**Comment by paagla123** [comment]
@paweenwich please ready https://blog.adstage.io/2013/11/26/tracking-conversions-with-google-analytics/ to learn about conversion using Google Analytics

**Comment by paagla123** [comment]
_From @daver78 on September 14, 2017 12:21_

@paagla123 and @paweenwich, I think neither of you has worked with Google Analytics in this kind of scenario, and thus the concepts seem foreign and complicated.  Let me clarify and try to simplify the mental picture you've created.
----------
Put simply, the question we're answering for the User (in the context of the Promote a Blog Post blueprint) is:

_Of the people who saw my blog post, how many of them went on to complete another action that is important to me?_

Put another way, this could be:

_Of the people who saw my blog post, how many of them went on to buy something from my site?_ or

_Of the people who saw my blog post, how many of them went on to download an eBook?_ or

_Of the people who saw my blog post, how many of them went on to start a free trial?_

This is the question we're answering by allowing the User to mark a conversion point (a page on their website that when reached, signifies an important step in the marketing funnel has been reached).

In order to know if someone has reached that page, we have many ways to accomplish this, but we're going to use Google Analytics for now.

Here's how:

1.  The User's website already has Google Analytics installed.  GA has 55% plus of the web tracking market so there's a good chance the marketer is already using them.

2.  In Step 1 of the Promote a Blog Post, we ask the User to (optionally) specify a page URL that when reached, signifies that the conversion point has been reached.  

3.  Now the User launches their campaign.  People start viewing their blog post.  A percentage of those people go to the next step (make a purchase, start a free trial, download something, whatever it is that's important to the marketer).

4.  On a very regular basis, we query Google Analytics (or configure GA to push data to us) and see who has reached the page the User marked in Step 2 above (the conversion URL).

5.  Now we can show the User their complete funnel: 

How many people they targeted -> how many opened the email -> how many clicked to the blog post -> and how many took the next important step (whatever it is that's important to the User).

**That's it.**

At this point, the only question in your mind should be: "_How do we match a visit to the conversion page (which we've pulled from GA) to a specific person in the Studio DB?_"

Fair question.  

What I suggested to Kushal is that we append something like `contactID=123456` to every link we have in the blueprint leading to the blog post.

Now, when we query GA to see how many people have visited that page, we can look for the `contactID` param, and get the `contactID` to do the match back to the Studio Contact.

Make sense?

Read this carefully and let me know what questions you have about the approach.

**Comment by paagla123** [comment]
@daver78 nice explanation thank you. I think I was on the wrong track. So we don;t need the goals

**Comment by paagla123** [comment]
_From @paweenwich on September 15, 2017 2:59_

@daver78 Thank you.   Now, I have clear scenario for the GA usage.  I have a few questions.   Assume the blog post blue print as explain.

1) So, we assume that the user's site already have GA code install in his/her account.  Is that possible that we can API the GA stats of other account.   If this can be done, do we need something like "userid" or "GA's id" from users in order to get the report?

2) Is utm and GA the same thing?  KD show use that we need to put GA code in html page.  How that relate to utm string that we ask user to put in?

3) What we should do with the "conversion" URL that user input.  If I am the marketer, I have did install GA.  If I know this much, I should be able to set the "conversion" URL myself.  (I hope it can be set in GA UI,  if not, please explain how conversion point work).  Then why I need this APP.


So, basically, I don't know how to do with all GA option that we ask user to put in.  If you can explain that, we will use xxxx  that user input to call API to get yyyy and do  zzzzz.   This will help a lot.

**Comment by paagla123** [comment]
_From @coa0329 on September 15, 2017 3:45_

https://developers.google.com/analytics/devguides/reporting/core/v4/quickstart/web-js

To use Analytics Reporting API. 
1 User must set the permission to use Analytics Reporting API in the account.
2 User give us client ID and view ID.

**Comment by paagla123** [comment]
_From @daver78 on September 15, 2017 3:48_

@paweenwich I will answer each of your questions in the same order:

1.  Yes, we should assume the User's site already has GA installed.  Yes, the GA API allows us to access the data.  Yes, we need information from the User in order to do this.

    To illustrate this, if you look at [the mockup](https://da-vinci-rules-daver78.c9users.io/httpdoc/daver/dv.php?page=myProfileSettings ) and then click on “3rd Party Product Linking” you'll see where we collect the User's `GA Tracking ID`.  

    However, I think we may need to do an oAuth connection to GA in order to receive permission to query the data, so the UI may need to change a bit to handle this.

2.  `UTM` is a URL pattern used by GA and other analytics tools to give the marketer the ability to measure traffic sources.  GA knows how to use UTM to display meaningful data within their reporting.  

    I'm not sure what KD showed you about putting GA code in HTML, but that's not required for this portion of the project.  As you stated in 1 (above), we assume the User has GA already installed on their site.

3.  We take the conversion URL that the User has entered, and query GA to see if anyone has reached that page.  If they have, then we know the blog post was responsible for driving the desired action.

@paweenwich when you say  _I don't know how to do with all GA option that we ask user to put in. If you can explain that, we will use xxxx that user input to call API to get yyyy and do zzzzz. This will help a lot._, does what I've written above clarify it for you, or are you asking me to specify which API calls I would make to retrieve the data?  I offered to do that for @paagla123 but he said he would figure it out with you.

@paagla123, I think you need to re-explain this portion in real-time with Kwang.  Perhaps you were on the wrong track last night, and my write-up above has clarified.

**Comment by paagla123** [comment]
_From @coa0329 on September 15, 2017 7:45_

I create the mockup to get the report from view "All Web Site Data" in account "StarbucksFirstTaste2017".

Go to http://davincistage.mindfireis.com/admin/HelloAnalyticsReport.html

Click Sign-in button. 
The google Sign-in will popup. 
Enter you google email and password. 
You will ask for permission. 
Click Allow.

You will see nothing , if you cannot access that view ["PERMISSION_DENIED"].
You will see the json of the report in textline, if you can access that view.

![135](https://user-images.githubusercontent.com/7914690/30471880-90faf6da-9a24-11e7-86ae-743ca7cd00b7.JPG)


We need client_id and VIEW_ID from user to access their account.

----------------------------


I create another mockup to get goal setting from my view on GA
Go to http://davincistage.mindfireis.com/admin/HelloAnalyticsUser.html
Wait until you see "get goal" button.
Click it, you will see my goal setting in the text line.
We need the following value to pass to the api : clientid, accountId, webPropertyId, profileId (view id) and goalId

----------------------------

**Comment by paagla123** [comment]
_From @daver78 on September 15, 2017 16:54_

@coa0329 Good progress.  

1.  Regarding the Google Sign button, we should place that in the proper section of Settings as I mention to @paweenwich above ([put it here](https://da-vinci-rules-daver78.c9users.io/httpdoc/daver/dv.php?page=myProfileSettings) in “3rd Party Product Linking” )

2.  What are you trying to do with the second mockup which retrieves the Goal Settings?  I believe I told @paagla123 that he was maybe on the wrong track when he spoke to you about that.  I believe he cleaned that up -- or no?

**Comment by paagla123** [comment]
_From @coa0329 on September 18, 2017 8:56_

@daver78 

The idea is to get all the goals to use as the option in dropdown for user to select.
We will show the report of that goal using GA api.

**Comment by paagla123** [comment]
_From @daver78 on September 18, 2017 14:14_

@coa0329 I think @paagla123 took you on the wrong path during the discussion of this part.  I thought he cleared that up.  @paagla123 Can you discuss tonight before @coa0329 does more work on this part?

**Comment by paagla123** [comment]
@coa0329 No need for Goals. Here's the discussion on Skype last Thursday on this topic:

Here're the API's https://developers.google.com/analytics/devguides/integrate/
Dave wants us to find an API that will retuen all URLs called and then search for a sring in the URL to compute how many people came from our email
So once you find API the rest should be simple

**Comment by paagla123** [comment]
_From @coa0329 on September 19, 2017 3:21_

![138](https://user-images.githubusercontent.com/7914690/30574367-a1558732-9d23-11e7-911c-1c44c1bf025a.JPG)


Do you mean we use GA api to get page visit then match it with Conversion URL user setting in each campaign ?

**Comment by paagla123** [comment]
@coa0329 That is correct. The conversion URL will have a particular string that we put there and search for it from the result of GA API calls.

**Comment by paagla123** [comment]
_From @coa0329 on September 19, 2017 7:2_

Goto http://davincistage.mindfireis.com/admin/HelloAnalyticsReport.html

![142](https://user-images.githubusercontent.com/7914690/30581468-6cbfb85e-9d4a-11e7-94e3-25aa9ca6389f.JPG)



I modify the code to get the pages from GA api.

So now we can use them to match with Conversion URL user setting, then create report.

**Comment by paagla123** [comment]
_From @daver78 on September 19, 2017 14:14_

@coa0329 OK, now I think you're on the right track.

@paagla123 I want to clarify a portion of last night's chat between you and @coa0329: I believe that passing `s=dv` is not enough; further, just passing a `campaignID` is not enough.  

**We need to automatically pass both a `campaignID` and `contactID` to the marketer's website.** 

Why?  We need the `campaignID` to associate the conversions to the right Da Vinci campaign, and we need the `contactID` to be able to identify which of the marketer's prospects/clients performed the action.

Make sense?

**Comment by paagla123** [comment]
_From @coa0329 on September 20, 2017 8:36_

We will add the following parameters "campaignID={{campaignID}}&contactID=##id##&source=EMAILX" to "call to action" link in all emails so we will have the information from GA.

This is the example link in email

http://davincistage.mindfireis.com/admin/GA_test2.php?campaignID=a45f6961a4d7f7783f9232b7dc9de381&contactID=1&source=EMAIL1&utm_medium=email&utm_source=Blog&utm_campaign=My+Blog&utm_term=Write

**Comment by paagla123** [comment]
_From @daver78 on September 20, 2017 15:18_

@coa0329 Good thinking on passing the email `source=EMAILX`.  Thanks for considering that important angle.  @paagla123 I assume you can see the validity of this param.

What's the next step?

**Comment by paagla123** [comment]
_From @coa0329 on September 25, 2017 12:1_

Working on it.

**Comment by paagla123** [comment]
@coa0329 remember to save the information about the person that converted+the source information as you recommended somewhere in the cDB. Also reminder to use contactID and not anything that is identifiable with the person.

**Comment by paagla123** [comment]
@coa0329 I have set up a campaign with conversion points for your to test. Its called Blog-0925-Google Analytics-Test .

**Comment by paagla123** [comment]
_From @coa0329 on September 26, 2017 3:26_

@paagla123 

cDB cannot do that, If you want to keep the person information we need to use mysql db.

We must create the daily cron to get data from GA to save in our db.

------------

What is GA Tracking Id you use ?

Please give the permission to tt994613@gmail.com to access it.

------------

**Comment by paagla123** [comment]
@coa0329 permission given. Why do you say cDB cannot do it?

**Comment by coa0329** [comment]
We can store data in cDB but it is very hard to get data to create report.

https://www.xaprb.com/blog/2010/09/07/a-gentle-introduction-to-couchdb-for-relational-practitioners/

CouchDB isn’t designed for arbitrary queries at runtime. You can only query one view, show function, or database at a time. You can’t do joins. You can’t do arbitrary GROUP BY and ORDER BY.

==================

**Comment by paagla123** [comment]
@coa0329 please use this in the GA Query Explorer to use as an example - that could be a starting point for the fingerprint - you will need to play with various options.
https://ga-dev-tools.appspot.com/query-explorer/?start-date=2017-10-15&end-date=2017-10-15&metrics=ga%3Ausers&dimensions=ga%3Abrowser%2Cga%3AbrowserVersion%2Cga%3AoperatingSystemVersion%2Cga%3Alatitude%2Cga%3Alongitude%2Cga%3AscreenResolution%2Cga%3AbrowserSize

**Comment by daver78** [comment]
@paagla123 I believe this is the order of tasks we discussed:

1- @paagla123 will experiment with various `env vars` to see if he can get a fairly close match between unique fingerprints and visitors to the MF website.  

One issue here is that Query Explorer limits you to 7 dims, so you'll need to find an alternate way to do this.

2- Once @paagla123 figures out which `env vars` to use, he'll let @coa0329 know which to use. If @paagla123 cannot figure this out, we may need to go a different route.

3- @paagla123 Assuming you pass step #2 above, I think you'll also need to write out, step by step, what should happen from a logical point of view so that @coa0329 knows how to write this code.  Recall that I believe you may want to use GA in the `handler.php` or microsite page, so that you're able to match fingerprints more easily.

**Comment by daver78** [comment]
@paagla123 and Dave decided to fall back to Plan B: Using our web-tracking method.  @paagla123 will take next steps with @javpowell and team.

**Comment by javpowell** [comment]
Maml: https://s3.amazonaws.com/documentsin/WebTracking.maml

Steps:
1)Publish MAML with a Schedule
2)Get Schedule ID the Schedule from Maml
3)Use this PHP file to get your Identifier necessary for the cloud connect callback URL passing account ID and Schedule ID: http://ccihelp.mindfireinc.com/cgi-bin/production/MindFire-QOS-Scripts/customDev/helpprod.php?accountId=[AccountID]&scheduleId=[ScheduleID]

(I have attached the source PHP in case you want to modify called helpprod.php
[helpprod.zip](https://github.com/coa0329/BhaangWeb/files/1416198/helpprod.zip)

It will return something like this:

{"identifier":"1_85319_636444520883916574","accountId":"1","outboundId":30932,"completeURL":"http://studio.afw.fxt.me/api/OutboundApp/AppCallback?serviceTypeId=2019&identifier=1_85319_636444520883916574&accountId=1&outboundId=30932"}

You will use identifier, accountid, and outboundid in step 4.

4)Paste In all Microsite pages to Set Cookie: <script type="text/javascript" src="http://track.mindfireinc.com/track.js?dom=[CookieName]&id=[Identifier]&aid=[AccountID]&oid=[OutboundID]&pid=##purl##"></script>

**Description of query params:**
Dom - Cookie name
Id - Identifier 
Aid - Account Id
Oid - Outbound Id
Pid - Purl (or any other dedupe variable)

5)To track activity on a third party website paste this code on all pages you want to track:

<script type="text/javascript" src="http://track.mindfireinc.com/track.js?dom=[CookieName]&track=true"></script>

**Comment by paagla123** [comment]
@daver78 I will discuss with @coa0329 when I am over in Bangkok

**Comment by daver78** [comment]
@paagla123 OK.  Are you clear with @javpowell how this solution would work?

**Comment by paagla123** [comment]
@daver78 I am clear at a 5,000 ft level. I will get clearer when I blunder into it!

**Comment by paagla123** [comment]
Closing this one and opening new issue for Web Tracking
